docs/proposals/20210210-insulate-users-from-kubeadm-API-changes.md

```diff
@@ -0,0 +1,226 @@
+
+This would allow to separate what users fill in in the KubeadmConfig/KubeadmControlPlane
+from which kubeadm API version Cluster API end up using in the bootstrap data.
+
+### Goals
+
+- Define a stop-gap for using most recent version of the kubeadm API in Cluster API
+  v1alpha3 - without introdicing any breaking change -.
+- Define how to stop to exposing the kubeadm v1betax types in the KubeadmConfig/KubeadmControlPlane
+  specs for v1alpha4.
+- Define how to use the right version of the kubeadm types generating our kubeadm yaml
+  file and when interacting with the kubeadm-config ConfigMap.
+- Ensure a clean and smooth v1alpha3 to v1aplha4 upgrade experience.
+
+### Non-Goals
+
+- Adding or removing fields in KubeadmConfig/KubeadmControlPlane spec.
+- Introduce behevioral changes in CABPK or KCP.
+
+### Future work
+
+- Evaluate improvements for the Cluster API owned version of KubeadmConfig/KubeadmControlPlane
+  specs types.
+- Make it possible for the Cluster API users to take benefit of changes introduced
+  in recent versions of the Kubeadm API.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a user, I want to user kubeadm providers for Cluster API (CABPK & KPC) in the same
+way across the entire spectrum of supported kubernetes versions.
+
+#### Story 2
+
+As the kubeadm tool, I want the kubeadm providers for Cluster API (CABPK & KPC) to
+use the latest kubeadm API supported by the target kubeadm/kubernetes version.
+
+### Requirements
+
+R1 - avoid breaking changes in v1alpha3
+R2 - ensure a clean v1alpha3 to v1alpha4 upgrade
+
+### Implementation Details/Notes/Constraints
+
+#### Background info about kubeadm API version
+
+kubeadm v1beta1 types:
+
+- Introduced in v1.13.
+- Deprecation cycle started with v1.17.
+- Removal scheduled for v1.20 but then postponed to v1.21.
+
+kubeadm v1beta2 types:
+
+- Introduced in v1.15.
+- Changes from the previous version are minimal:
+  - Support for IgnorePreflightErrors into the NodeRegistrationOptions.
+  - Support for CertificateKey into InitConfiguration and JoinConfiguration
+    configuration (not relevant for Cluster API because it is not using the
+    automatic certificate copy feature).
+  - improved serialization support (in practice a set of `omitempty` fixes).
+
+#### Background info about kubeadm types into the KubeadmConfig/KubeadmControlPlane specs
+
+Given the fact that importing kubead (which is part of the Kubernetes codebase)
+in Cluster API is impratical, Cluster API hosts a mirror of kubeadm API types.
+
+kubeadm v1beta1 mirror-types:
+
+- Hosted in `bootstrap/kubeadm/types/v1beta1`.
+- Diverged from the original v1beta1 types for better CRD support (in practice
+  a set of `+optional` fixes, few `omitempty` differences).
+
+kubeadm v1beta2 mirror-types:
+
+- Hosted in `bootstrap/kubeadm/types/v1beta2`.
+- Currently, not used in the Cluster API codebase.
+- Does not include changes for better CRD support introduced in kubeadm v1beta1
+  mirror-types.
```

Comment:
`+optional` is not present in v1beta2 of kubeadm.

does this mean if CAP* moves to v1beta2 it has to apply the same forked changes?

do we want to apply these changes to a future kubeadm v1beta3?

Reply:
> does this mean if CAP* moves to v1beta2 it has to apply the same forked changes?

Yes, I would like to this consistent

> do we want to apply these changes to a future kubeadm v1beta3?

IMO yes, this is basically part of making kubeadm types to be CRDs

Reply:
https://github.com/kubernetes-sigs/cluster-api/pull/4204 aligned the v1beta1 and v1beta2 forked types.